search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,447 @@
+  state->invocation = invocation;
+
+  return state;
+}
+
+static void
+article_metadata_query_state_free (ArticleMetadataQueryState *state)
+{
+  g_free (state);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ArticleMetadataQueryState,
+                               article_metadata_query_state_free);
+
+typedef struct GSList ObjectSList;
+
+static void
+object_slist_free (ObjectSList *slist)
+{
+  g_slist_free_full ((GSList *) slist, g_object_unref);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ObjectSList, object_slist_free);
+
+static void
+add_key_value_pair_to_variant (GVariantBuilder *builder,
+                               const char      *key,
+                               GVariant        *value)
+{
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("{sv}"));
+  g_variant_builder_add (builder, "s", key);
+  g_variant_builder_add (builder, "v", value);
+  g_variant_builder_close (builder);
+}
+
+static inline gchar *
+underscorify (const gchar *string)
+{
+  gchar *underscored = strdup (string);
+  gchar *c = underscored;
+  do
+    {
+      if (*c == '-')
+        *c = '_';
+    }
+  while (*c++);
+
+  return underscored;
+}
+
+static void
+maybe_add_key_value_pair_from_model_to_variant (EkncContentObjectModel *model,
+                                                GVariantBuilder        *builder,
+                                                const char             *key,
+                                                const GVariantType     *expected_type)
+{
+  g_auto(GValue) value = G_VALUE_INIT;
+  g_autofree gchar *underscore_key = underscorify (key);
+  GParamSpec *pspec = g_object_class_find_property (G_OBJECT_GET_CLASS (model),
+                                                    key);
+
+  if (pspec == NULL)
+    return;
+
+  g_value_init (&value, pspec->value_type);
+  g_object_get_property (G_OBJECT (model), key, &value);
+  add_key_value_pair_to_variant (builder,
+                                 underscore_key,
+                                 g_dbus_gvalue_to_gvariant (&value,
+                                                            expected_type));
+}
+
+static void
+on_received_query_results (GObject      *source,
+                           GAsyncResult *result,
+                           gpointer      user_data)
+{
+  EkncEngine *engine = EKNC_ENGINE (source);
+  g_autoptr(ArticleMetadataQueryState) state = user_data;
+  g_autoptr(ObjectSList) models = NULL;
+  g_autoptr(ObjectSList) shards = NULL;
+  g_auto(GStrv) shards_strv = NULL;
+  g_auto(GVariantBuilder) builder;
+  g_autoptr(GError) error = NULL;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+
+  g_application_release (g_application_get_default ());
+
+  if (!models_and_shards_for_result (engine,
+                                     state->provider->application_id,
+                                     result,
+                                     (GSList **) &models,
+                                     (GSList **) &shards,
```

Comment:
Having to cast ObjectSList back to GSList like this all the time just so that we could get autoptr functionality seems rather daft, but I can't think of a better way to do ti

Reply:
It might work if you do `typedef GSList ObjectSList` without the `struct`? Otherwise I'd say it's too cumbersome if it's only for the benefit of this one function, just use `goto out`. Or use `__attribute__((cleanup(object_slist_free)))` directly.

Reply:
My preference is probably to use the first option in that case. The point of `autofree` was to avoid having to use `goto` and the latter approach is too cumbersome IMO.